### PR TITLE
Add BBE for clone builtin method

### PR DESCRIPTION
**Pull request by gimantha**
## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: examples/clone/clone.bal [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+public type Address record {
+    string country;
+    string state;
+    string city;
+    string street;
+};
+
+// The returned tuple will be later used by ballerina tests
+public function main() returns (Person, Person, string)? {
+    // Create a value of Address
```
Comment (samgnaniah) [inline examples/clone/clone.bal]: Shall we say something like "Specify values for the Address record."?

File: examples/clone/clone.bal [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    Address address = {
+        country : "USA",
+        state: "NC",
+        city: "Raleigh",
+        street: "Daniels St"
+    };
+
+    // Create a value of type Person
```
Comment (samgnaniah) [inline examples/clone/clone.bal]: Same comment as above is applicable.

File: examples/clone/clone.bal [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        io:println("Cloned value: ", result);
+        string refCheck = "";
+        // Check if the reference of both values is not the same.
+        if (result !== person) {
+            refCheck = "Source and Clone are at two different memory locations";
+            io:println(refCheck);
+        }
+        return (person, result, refCheck);
+
+     // If the result is an error, prints out the error and return ()
```
Comment (samgnaniah) [inline examples/clone/clone.bal]: Change to "If the result is an error, the error is printed out and () is returned."

File: examples/clone/tests/clone_test.bal [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import ballerina/test;
+
+@test:Config
+function testFunc() {
+    // Invoke the main function.
+    (Person, Person, string)? result = main();
+
+    if (result is (Person, Person, string)) {
```
Comment (samgnaniah) [inline examples/clone/tests/clone_test.bal]: Shall we add a comment to this as well?

File: examples/clone/clone.bal [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    // Create a value of type Person
+    Person person = {
+        name: "Alex",
+        age: 24,
+        married: false,
+        salary: 8000.0,
+        address: address
+    };
+
+    // Note that we have assigned the person value to any typed variable. So the type of the value has to be determined
```
Comment (samgnaniah) [inline examples/clone/clone.bal]: Shall we shorten this and just say "Assign the Person value..."

File: examples/clone/clone.bal [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    // in runtime.
+    any anyValue = person;
+
+    // Invoke the clone builtin method. The return type is either Person or error. Note that the error is returned if
+    // the value being cloned is not of anydata type. If the type of the value being cloned can be determined in
```
Comment (samgnaniah) [inline examples/clone/clone.bal]: Add "anydata" between single back ticks. Apply to all such instances and similar values in the example.

File: examples/clone/clone.bal [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    // compile time, return value will be exactly of the type of the value being cloned.
+    // e.g. Person result = person.clone();
+    // In above example, 'person' is cloned instead of 'anyValue'. In compile time itself, we know the return type
+    // should be 'Person'
+    var result = anyValue.clone();
+
+    // Type of any variable is only determined in runtime. Therefore we have to check if the returned type is Person
+    // or an error. If we would have cloned 'person' value, we shouldnt be checking for the type of the return value.
```
Comment (samgnaniah) [inline examples/clone/clone.bal]: Shall we change this to something like the following: "If we cloned the 'person' value, we do not need to check the type of the return value."

File: examples/clone/clone.bal [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+};
+
+public type Address record {
+    string country;
+    string state;
+    string city;
+    string street;
+};
+
+// The returned tuple will be later used by ballerina tests
```
[the remark this answers is not in the document]
Comment (gimantha) [reply examples/clone/clone.bal]: Fixed.

File: examples/clone/clone.bal (hunk 2 above)
Comment (gimantha) [reply examples/clone/clone.bal]: Fixed.

File: examples/clone/clone.bal (hunk 1 above)
Comment (gimantha) [reply examples/clone/clone.bal]: Fixed.

File: examples/clone/clone.bal (hunk 5 above)
Comment (gimantha) [reply examples/clone/clone.bal]: I changed it to "// Assign the person value to anyValue. So the type of the value has to be determined in runtime.
    ". Can you please review. I think we have to specifically say why we need to assign the Person to anyValue

File: examples/clone/clone.bal (hunk 6 above)
Comment (gimantha) [reply examples/clone/clone.bal]: Fixed

File: examples/clone/clone.bal (hunk 7 above)
Comment (gimantha) [reply examples/clone/clone.bal]: fixed

File: examples/clone/tests/clone_test.bal (hunk 4 above)
Comment (gimantha) [reply examples/clone/tests/clone_test.bal]: Added

File: examples/clone/clone.description [hunk 9]
```diff
@@ -0,0 +1,4 @@
+// Clone built-in method is used to make a copy of any anydata value.
+// Clone is defined on all structured types but return an error if value is not anydata.
+// The resultant copy will be a deep copy
+// It preserves cycles, it preserves the storage type and it preserves immutability
```
[the remark this answers is not in the document]
Comment (gimantha) [reply examples/clone/clone.description]: Added

File: examples/clone/clone.description [hunk 10]
```diff
@@ -0,0 +1,4 @@
+// Clone built-in method is used to make a copy of any anydata value.
+// Clone is defined on all structured types but return an error if value is not anydata.
+// The resultant copy will be a deep copy
```
[the remark this answers is not in the document]
Comment (gimantha) [reply examples/clone/clone.description]: Fixed

File: examples/clone/clone.description [hunk 11]
```diff
@@ -0,0 +1,4 @@
+// Clone built-in method is used to make a copy of any anydata value.
+// Clone is defined on all structured types but return an error if value is not anydata.
```
[the remark this answers is not in the document]
Comment (gimantha) [reply examples/clone/clone.description]: Fixed

File: examples/clone/clone.description [hunk 12]
```diff
@@ -0,0 +1,4 @@
+// Clone built-in method is used to make a copy of any anydata value.
```
[the remark this answers is not in the document]
Comment (gimantha) [reply examples/clone/clone.description]: Fixed.

File: examples/clone/clone.bal (hunk 3 above)
Comment (gimantha) [reply examples/clone/clone.bal]: Fixed.

File: examples/clone/clone.bal (hunk 5 above)
Comment (samgnaniah) [reply examples/clone/clone.bal]: Yes, this is fine.

**Pull request merged**